src/GenericContainer/include/BipedalLocomotion/GenericContainer/Vector.h

```diff
@@ -820,5 +873,156 @@ typename Vector<const typename container_data<Class>::type>::eigen_map_const_typ
+     * @param other The input vector from which the context is taken
+     * In principle, Ref should be the reference of a Vector<T> which should remain alive while
+     * Ref is alive. On the other hand, Vector<T> is only a pointer to some data which does not own.
+     * Hence, Ref can remain alive even if the Vector<T> is deleted, provided that the original container
+     * stays alive.
+     */
+    Ref(BipedalLocomotion::GenericContainer::Vector<T>&& other)
+    {
+        m_span = other.m_span;
+        m_resizeLambda = other.m_resizeLambda;
+    }
+
+    /**
+     * Constructor from another container.
+     * This is used if:
+     * - the input container is not a GenericContainer::Vector, to avoid ambiguities with other constructors
+     * - the input container is not a string. This allows using Ref and string with overloaded methods.
+     * - a GenericContainer::Vector<T> can be constructed from the Container
+     * - T is not const
+     * - the input container is not const.
+     */
+    template <class Vector, typename = typename std::enable_if<!GenericContainer::is_vector<Vector>::value &&
+                                                               !std::is_same<Vector, std::string>::value &&
+                                                               GenericContainer::is_vector_constructible<Vector>::value &&
+                                                               !std::is_const_v<T> &&
+                                                               !is_container_const<Vector>::value>::type>
+    Ref(Vector& input)
+    {
+        if constexpr (is_span_constructible<Vector>::value)
+        {
+            m_span = iDynTree::make_span(input);
+        } else
+        {
+            m_span = iDynTree::make_span(input.data(), input.size());
+        }
+
+        if constexpr (BipedalLocomotion::is_resizable<Vector>::value)
+        {
+            m_resizeLambda = DefaultVectorResizer(input);
+        }
+        else
+        {
+            iDynTree::Span<T> copiedSpan = m_span;
+            m_resizeLambda = [copiedSpan](index_type size){unused(size); return copiedSpan;};
+        }
+
+    }
+
+    /**
+     * Constructor from another container.
+     * This is used if:
+     * - the input container is not a GenericContainer::Vector, to avoid ambiguities with other constructors
+     * - the input container is not a string. This allows using Ref and string with overloaded methods.
+     * - a GenericContainer::Vector<T> can be constructed from the Container
+     * - T is const.
+     */
+    template <class Vector, typename = typename std::enable_if<!GenericContainer::is_vector<Vector>::value &&
+                                                               !std::is_same<Vector, std::string>::value &&
+                                                               GenericContainer::is_vector_constructible<Vector>::value &&
+                                                               std::is_const_v<T>>::type>
```

Comment:
Just a curiosity. Can this be SFINAE construct be simplified using an `if constexpr` in this specific scenario? If not, could you tell me why?

Reply:
SFINAE makes the function/method below disappear. This allows the compiler to choose one version or the other. ``if constexpr`` instead is used *inside* the function. Hence, you can use it to compile a method in two different ways while maintaining the same signature. Rule of thumb:
- use SFINAE to drive the compiler toward the correct method, when different methods have ambiguous signatures
- use ``if constexpr`` to conditionally compile some pieces of code inside a method depending on some condition known at compile time.

With ``c++20`` SFINAE may be substituted with a [``concept``](https://en.cppreference.com/w/cpp/language/constraints)

Reply:
Understood! Thanks for the pointers!